### PR TITLE
spack containerize: permit to customize the base images

### LINKED ISSUE #14802: Additional containerize spec file customization
The [containerize feature](https://github.com/spack/spack/pull/14202) should allow more customization of the container spec file.

### Rationale

Consider the use case of an application container to be run multi-node over InfiniBand.  The MPI library needs to be built with fabric support.  E.g.,

```
spack:
  specs:
  - gromacs@2019.4 build_type=Release
  - openmpi@3.1.4 fabrics=verbs
  - fftw precision=float
  packages:
    all:
      target: [broadwell]

  container:
    format: docker

    base:
      image: "ubuntu:16.04"
      spack: develop

    strip: true

    os_packages:
    - libgomp1
```

This container build will fail because the IB verbs packages are not present in the container base image.

### Description

The set of base images is highly restricted, so the option of using a custom base image with the IB libraries already included is not allowed.

The `os_packages` section applies only to the runtime stage of the spec, not the build stage.  Regardless, the packages are installed after spack, not before as would be necessary in this case.

The `extra_instructions` `build` section is executed after spack, so it's also not an option.

Based on the failed approaches, there are 3, not mutually exclusive, potential containerize enhancements.

1. Remove the base image restriction.  Perhaps display a warning if a custom base image is selected, but don't disallow it outright.

2. Extend the `os_packages` section to include `build` and `runtime` sections.  E.g.,
```
  os_packages:
    build: libibverbs-dev
    runtime: libibverbs
```

The behavior should also be changed to install the packages before spack, not after.

3. Add a new `extra_instructions` section that is inserted prior to spack, e.g., `prebuild` or `setup`.
```
  extra_instructions:
    prebuild: RUN apt-get update -y && apt-get install libibverbs-dev 
```

My suggestion would be 2 to address this particular use case, but 3 would be a more general, if less user friendly, solution.

### PR DESCRIPTION
**Pull request by alalazo**
closes #14802 
closes #14879

This PR reworks a few attributes in the `container` subsection of `spack.yaml` to permit the injection of custom base images both in the build and final stage. In more detail, users can still specify the base operating system and Spack version they want to use:
```yaml
spack:
  container:
    images:
      os: ubuntu:18.04
      spack: develop
```
in which case the generated recipe will use one of the Spack images built on Docker Hub for the build stage and the base OS image in the final stage or alternatively they can specify explicitly the two base images:
```yaml
spack:
  container:
    images:
      build: spack/ubuntu-bionic:latest
      final: ubuntu:18.04
```
and it will be up to them to ensure their consistency. The ability to install OS packages in both the `build` and `final` stage has been retained. Handles to avoid an update of the available system packages have been added to the configuration to facilitate the generation of recipes permitting deterministic builds. 

~Following the discussion in #14917 this PR takes at the moment a minimalist approach and thus removed the possibility to install OS packages in the generated image (if any is needed it should be the custom base image that takes care of that). In case this seems unwanted it will be sufficient to drop commit ed71483 to make this PR compatible with #14917 and #14879.~

### CODE REVIEW

**Comment by alalazo** [comment]
@ChristianTackeGSI @samcmill

**Comment by samcmill** [comment]
My 2 cents is to retain the ability to install OS packages in the image.

The rationale is to have something in-between the canned images (zero effort but zero control) and custom images (lots of control but potentially lots of effort).

File: lib/spack/docs/containers.rst
Comment (hartzell) [inline lib/spack/docs/containers.rst]: It would be helpful to add some documentation about what, if any, things MUST be present in the build and final images.  At a minimum, point out that `spack containerize` makes *some* assumptions and refer/link to the relevant Dockerfiles as an existence proof of what is known to work.

File: lib/spack/docs/containers.rst
Comment (hartzell) [inline lib/spack/docs/containers.rst:299]: I won't miss this feature, but I see that @samcmill likes it.  If it goes back in it would be nice to be able to choose whether to update or not (I know, it's less simple but it's more reproducible, so choose your poison...).

Comment (alalazo) [reply lib/spack/docs/containers.rst:299]: > If it goes back in it would be nice to be able to choose whether to update

Sure. That wouldn't be a big change. In case we also need a way to let the user specify which system package manager we are using - in case of a custom image that would be the simplest and more effective way to generate the lines we need in the recipe. So something like:
```yaml
os_packages:
  command: yum
  update: False
  build:
  - libibverbs-dev
  final:
  - libibverbs
```
If we are using the images distributed on Docker Hub then the system package manager will be deduced from the base image like it is now.

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: Eheh, to clarify: this is not the final state of the docs and I completely agree with you on the need to specify better the requirements of the two images. I'm just trying to wait to see where we are going before extending this part with more subsections.

File: lib/spack/docs/containers.rst
Comment (hartzell) [reply lib/spack/docs/containers.rst:299]: Sounds great!

File: lib/spack/docs/containers.rst
Comment (hartzell) [reply lib/spack/docs/containers.rst]: 👍

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst:299]: Done.

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: I added a few bits to the docs. Let me know what you think.

**Comment by reidpr** [comment]
Thanks for your hard work on this PR.

May I make a friendly request? Can everyone please use gender-neutral language when describing users? For example, instead of “a user can still specify ... Spack version he wants to use”, write “a user can still specify ... Spack version they want to use”.

Among lots of other good reasons, this would be in keeping with the Spack COC, e.g. the admonition to use “inclusive language”.

**Comment by alalazo** [comment]
@reidpr Edit done. I'll have a look at the docs later and adjust it too if I find anything. Thanks.

**Comment by victorusu** [comment]
This is great! please merge!

**Comment by alalazo** [comment]
@hartzell Do you have further requests wrt to docs or anything else?

**Comment by alalazo** [comment]
@svenevs I thought that given the positive comment in Spack's slack on `spack containerize` you might be interested in this PR :slightly_smiling_face:

File: lib/spack/docs/containers.rst
Comment (paulbry) [inline lib/spack/docs/containers.rst:185]: @alalazo First, I'm really excited about this. When just following along and copying from this example I ran into the following:

```
$ spack containerize > Dockerfile
...
ruamel.yaml.scanner.ScannerError: while scanning a simple key
  in "/Users/pbryant/Development/10-Containers/spack-containerize/gromacs/spack.yaml", line 27, column 1
could not find expected ':'
  in "/Users/pbryant/Development/10-Containers/spack-containerize/gromacs/spack.yaml", line 28, column 1
```

I was edit able to edit the file in order to get it working but wanted to bring it up.

Edit: Apologies, referring to `RUN echo ...`

Comment (alalazo) [reply lib/spack/docs/containers.rst:185]: @paulbry Sorry, I don't understand the edit. Is there an issue with any example in this PR?

Comment (paulbry) [reply lib/spack/docs/containers.rst:185]: @alalazo  It's my fault, I commented on the wrong lines of code and should have corrected it. The section that caused an issue was:

```
       extra_instructions:
         final: |
   RUN echo 'export PS1="\[$(tput bold)\]\[$(tput setaf 1)\][gromacs]\[$(tput setaf 2)\]\u\[$(tput sgr0)\]:\w $ \[$(tput sgr0)\]"' >> ~/.bashrc
```

Simply copying/pasting that section is what caused the error I provided above. I just indented the `RUN echo ...` line to correct this.

File: share/spack/templates/container/singularity.def
Comment (scheibelp) [inline share/spack/templates/container/singularity.def]: Does the template format require that there is no spaces before the directive? In other words, is it possible to write

```
  {% if os_package_update %}
```

(i.e. there are two spaces in front)

? If so, can you update the definition file? It would make it easier to read when certain conditions apply.

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [inline lib/spack/spack/schema/container.py:65]: Could `snap` or `flatpak` also work? In other words, why constrain this?

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [inline lib/spack/spack/schema/container.py:61]: Update the comment

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [inline lib/spack/spack/schema/container.py:11]: (main concern) If I understand correctly, "build" and "final" are used in the same way as "os" and "spack". They have different names to allow you to distinguish the constrained and unconstrained case in the schema. I'm inclined to say it would be clearer to:

* add a separate section which lists valid build/final images when a user wants to constrain them
  * ... and add these values to the default config (in `etc/spack`)
* add a variable that determines whether Spack should ensure that a specified build/final image is a preapproved one
  * ... and default that variable to true

That way the user doesn't have to choose different config variables to achieve the same effect.

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [inline lib/spack/spack/schema/container.py:68]: (opinion) I think "runtime" is clearer than "final" (for example the explanatory comment in the code above this talks about "packages that are needed at runtime"). If the term "final" comes up frequently in the container community and has a well-understood meaning, then I think "final" is fine.

File: lib/spack/spack/schema/container.py
Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: > If I understand correctly, "build" and "final" are used in the same way as "os" and "spack". They have different names to allow you to distinguish the constrained and unconstrained case in the schema. I'm inclined to say it would be clearer to [ ... ]

I don't really agree with this. The "os" + "spack" configuration is basically a simplified way to configure image generation if a user is fine with starting from our base images on Dockerhub. This is achieved by:

1. Constraining the images (as you noted)
2. Ensuring consistency between base images in the build and final stage

Consider for instance:
```yaml
images:
  os: ubuntu:18.04
  spack: 0.14.1
```
compared to what is my understanding of the proposal above:
```yaml
constrained: True
images:
  build: spack/ubuntu-bionic:0.14.1
  final: ubuntu:18.04
```
In the second case we require to the user to be needlessly more verbose and duplicate information (`ubuntu-bionic` and `ubuntu:18.04`), in the first case we just ask the minimal set of information we need to proceed with generating the container recipe. 

Overall I don't think having something in-between what we have now and removing the simplified mode is worthwhile, as:

1. We wouldn't be that much of an help in simplifying configuration to the user
2. We need to maintain nonetheless all the code that enforces constraints

My personal opinion would be to maintain this as is, but if people feel strongly I would be for just removing the "os" + "spack" mode and have no "simplified mode".

> I have a few minor requests but overall I think that "build"/"final" and "os"/"spack" are duplicate ways to specify the same thing, which I think is confusing.

Can you elaborate on what can be confusing / what do you think will annoy users with this setup? Asking because the discussion here involves a few power users and this is the first time the argument is raised, so I'd like to understand better the concerns.

File: lib/spack/spack/schema/container.py
Comment (alalazo) [reply lib/spack/spack/schema/container.py:68]: The word "final" is somewhat used when talking of multi-stage builds (as in the "final image" etc. see [here](https://docs.docker.com/develop/develop-images/multistage-build/)). On the other hand the word "runtime" is used a lot when talking about container technologies with a complete different meaning (it's the mechanism by which a container image is run on the system), so I would avoid that particular term and maintain the current one.

File: share/spack/templates/container/singularity.def
Comment (alalazo) [reply share/spack/templates/container/singularity.def]: > Does the template format require that there is no spaces before the directive?

The spaces before the directives are rendered, at least by default. Adding the spaces there would result in a less readable file once the template is instanced. I can check if `jinja2` has options to interpret a line with a directive as something that shouldn't be rendered.

File: lib/spack/spack/schema/container.py
Comment (alalazo) [reply lib/spack/spack/schema/container.py:65]: Each package manager has ways to update the list of packages that can be installed, install them and remove all the data that is not anymore needed once the installation finishes. This is constrained because Spack currently knows only how to deal with `apt` and `yum` (but can be extended later).

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [reply lib/spack/spack/schema/container.py:11]: > 1. We wouldn't be that much of an help in simplifying configuration to the user
> Asking because the discussion here involves a few power users

Introducing two different ways of specifying the same thing is confusing to regular users. The variables have overlapping meaning but change behavior based on which ones you choose. That's easy to mix up.

> 2. We need to maintain nonetheless all the code that enforces constraints

It took me a while to realize how much these variables overlap when I read through the code, so I also think this would improve readability. 

I think we should try to figure out how to maintain a single set of config variables to specify containers and then add logic to improve things for power users. For example:

```
constrained: True
images:
  build: ubuntu:18.04
  final: 0.14.1
```

Is there any way Spack could infer the values of `spack/ubuntu-bionic:0.14.1`/`ubuntu:18.04` from `ubuntu:18.04`/`0.14.1`?

The `constrained` config variable could also move "up" and apply simultaneously to all specs to reduce tedium (although in that case it should have a more-descriptive name).

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: > Introducing two different ways of specifying the same thing is confusing to regular users. The variables have overlapping meaning but change behavior based on which ones you choose. That's easy to mix up.

I tried to solve this issue by adding explicit docs. See 63333ba530d6c8a72ab7578a795f75ba4e1c593e. Maybe that should be better worded?

> I think we should try to figure out how to maintain a single set of config variables to specify containers and then add logic to improve things for power users.

What we might do is the opposite of your example + avoid the `constrained` variable:
```yaml
images:
  build: 0.14.1
  final: ubuntu:18.04
```
Here the idea is that if `build` has a certain format it is interpreted as a Spack version, otherwise it's interpreted as an image on Docker hub. Personally I'd still prefer to have different labels since they make explicit this change, but let me know if that can be a good compromise.

> It took me a while to realize how much these variables overlap when I read through the code, so I also think this would improve readability.

Hmm, I don't see that in the sense that inner logic would remain the same but exposed to user in a different way. Maybe I just don't understand what you mean here.

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: @scheibelp I got some additional feedback for the current interface from a user (reporting it here to add on the discussion):
1. The simplified interface with `os` and `spack` is seen as intuitive (i.e. understandable from just reading the YAML) so those names should not be abandoned
2. The interface with `build` and `final` is more confusing than the simplified interface and it's difficult to understand the meaning of those attributes if you don't read the documentation carefully
3. The proposal to maintain only `build` and `final` (or only `os` and `spack`) and switch their meaning with a Boolean is equally confusing

The proposal that was raised after a bit of discussion is to:
- Maintain the two sets of variables ("`os`+`spack`" and "`build` + `final`") with the meaning they have now
- Add a Boolean variable under `images` named `simplified` that defaults to `True` and controls whether we want to use one set of variables or the other.

In that way the simplified interface will look the same as it is now, but the switch of meaning requires an explicit step by the user. For instance:
```yaml
images:
  os: ubuntu:18.04
  spack: 0.14.1
```
would be an example of the simplified interface, while
```yaml
images:
  simplified: False
  build: alalazo/cuda-10.1-ubuntu18.04:latest
  final: nvidia/cuda:10.1-base-ubuntu18.04
```
would be how custom images can be chosen.

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: Would the proposal above to have an explicit switch for the interface be fine with you?

Comment (scheibelp) [reply lib/spack/spack/schema/container.py:11]: I think we agreed that if more users think these separate configuration variables are useful (i.e. os/spack for the pre-made case and build/final for the custom) case, then that's what we support. If you can find folks to weigh in on it then we can make progress on that; if not then we can keep discussing it.

If we end up maintaining both os/spack and build/final, I think there needs to be some work to improve how they are handled in the code to make it obvious to future maintainers what they are intended for.

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: @hartzell @victorusu @paulbry Wondering if you have an opinion on the discussion above. The main point is if you find confusing the current interface with os/spack and a build/final to select between the two ways to specify images and if you have suggestions on how it should look like from user's perspective.

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: Pinging also @nazavode since I already bothered him for https://github.com/spack/spack/pull/15028#discussion_r403925148

Comment (nazavode) [reply lib/spack/spack/schema/container.py:11]: I think @alalazo summarized well my points about the UI in https://github.com/spack/spack/pull/15028#discussion_r403925148. TL;DR, as a user:

1. _simplified UI_ should be the default, be intuitive, work out of the box, straightforward;
2. _advanced UI_ should be inconvenient, require user intervention and mutually exclusive.

The proposed boolean switch checks all the marks imo, I could reach out to some users I'm currently working with to ask for their impressions if it could be beneficial.

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: @scheibelp Anything I can do to help this discussion on the YAML format moving forward? Is https://github.com/spack/spack/pull/15028#discussion_r403925148 fine? Should we seek more opinions on Slack?

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: I've been reading through discussion like https://github.com/spack/spack/pull/15028#discussion_r403925148 again and I think this schema could be workable but I think these docs need to be improved. For example:

> The ``images`` attribute in the ``container`` section has two modes of operation to select a proper base image for both the ``build`` and``final`` stage.

This is confusing because this sentence specifies that there are two modes, but then only specifies the variables for one of those modes. It implies that these variables are used in both cases.

Overall I think the description should do the following

* specify that their are two modes
* specify the variables used in each mode
* include the exact yaml for each example in this section (you avoid this for os/spack because it appears in a larger example above, but that example doesn't specifically address these details so I think they should specifically be called out in this section)
* specify at a high level how spack operates in each mode
  * e.g. `in os/spack mode, the user specifies a Spack version and OS. Spack has pre-constructed images for the OS with that Spack version pre-installed; Spack only supports a limited number of versions/operating systems in this way, so in this case the OS used is constrained. Spack performs the build on an image of the OS with Spack, then installs files to a corresponding base image of the OS (i.e. without Spack installed).`
  * e.g. `in build/final mode the user can specify any image to perform the Spack build on and any image to install the resulting files to. In this mode the user is responsible for creating the images and managing all of their details (e.g. that the desired Spack version is provided).`

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst:185]: Should be fixed now, thanks

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: @scheibelp I reworked this section completely, let me know what you think of the docs now. If this seems the right way to proceed I'd be happy to add a Boolean (as we discussed off-line):
```yaml
container:
  images:
    advanced: True
    build: myrepo/myimage:latest
    final: ubuntu:18.04
```
to explicitly activate the advanced mode.

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: I think when you are using the terms "build" and "final" here, you are not referring to the syntax introduced in this PR, so I think this should be rephrased. For example:

> The ``images`` subsection is used to select the docker image where Spack builds the package and the image where the built package is installed

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: @scheibelp ping

File: lib/spack/spack/schema/container.py
Comment (hartzell) [reply lib/spack/spack/schema/container.py:11]: At the risk of suggesting something that's none of the above, perhaps we could require that one of two section be specified, either

```yaml
stock_images:
  os: ubuntu:18.04
  spack: 0.14.1
```

or 

```yaml
custom_images:
  build: my_build_image
  final: my_final_image
```

It took me a while to figure out that the `spack` key was a [commit-ish](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefcommit-ishacommit-ishalsocommittish) within the Spack repository, presumably that'll be documented (assuming of course that a tag/branch/commit is valid there.

Comment (alalazo) [reply lib/spack/spack/schema/container.py:11]: > presumably that'll be documented (assuming of course that a tag/branch/commit is valid there.

@hartzell There's doc already for that in this PR, see [here](https://github.com/spack/spack/blob/d6acdd20ffe213e48e13d017b77a21058e18931f/lib/spack/docs/containers.rst). The part that explains possible values for `os` and `spack` is:

> Any combination of OS and Spack version that can be matched to an image in the Spack repository on Dockerhub is admissible as a value. By employing this mode the user instructs Spack to generate a recipe that will start from one of the official Spack images in the build stage and from the base OS image in the final stag

**Comment by hartzell** [comment]
FWIW, I just used a copy of this from a few days ago (d6acdd20f, before your push today) to build an image using my own custom build and base images.  The build image was the spack cento7 builder w/ gcc@8.4.0 pre{installed,configured} and the base image was standard centos7 image.

The resulting application works!

**Comment by hartzell** [comment]
Love to see it land.

File: share/spack/templates/container/singularity.def
Comment (scheibelp) [reply share/spack/templates/container/singularity.def]: Now that I think on this more, I'm wondering if it actually does make it less readable: wouldn't this just be some extra lines in the file?

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [reply lib/spack/spack/schema/container.py:65]: (no change required) That's true, so I'd say it might make sense to frame the API in terms of these commands rather than the package manager. And if we're going to infer commands from a package manager wouldn't it make it easier to do it based on the OS instead? In this case, I'm more interested if what I'm saying makes sense as a future refactor, vs. changing this in this PR.

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [reply lib/spack/spack/schema/container.py:11]: Perhaps you could clarify this further with an example:

> For example to use the "spack/ubuntu-bionic:0.14" image, you can set...

**Comment by hartzell** [comment]
This might be off tangent, but since I mentioned it in my experiment above....

I "want" to use CentOS and/but I need to use newer compilers.  My solution to that is to include gcc@8.4.0 in the `build` image, but I discovered that I *also* need to include it in the `final` image.  If I remove it, various libraries end up missing.  E.g. bcl2fastq2 depends on boost which ends up needing the libgomp in the gcc install tree.  (Oddly enough, libgomp is included in the example, but even if I use the yum version I end up missing other things).

So, in spite of vary carefully doing a multistage build, I end up with a fairly large image (gcc is about 1.2GB).  I could thin it by carefully trimming most of the gcc install tree (leaving just lib and lib64 might work), but that seems fragile.

Is there something else I could/should be doing (don't say CentOS8...), or is this a fact of life?

**Comment by hartzell** [comment]
Hi All.  Any movement on this?

**Comment by alalazo** [comment]
@hartzell I'm keeping this PR rebased semi-frequently. Will try to have it merged for the next major release.

**Comment by hartzell** [comment]
Great.  I'm currently running `spack containerize` and piping the output through `sed`, so I'll look forward to this landing.

FWIW, my custom builder pre-builds gcc while my custom final needs to include our internal root CA certs.

Is there a timeframe for the next major release?

**Comment by alalazo** [comment]
> Is there a timeframe for the next major release?

Before SC next month :slightly_smiling_face:

**Comment by hartzell** [comment]
Woo hoo!   **Thanks!**

File: share/spack/templates/container/singularity.def
Comment (alalazo) [reply share/spack/templates/container/singularity.def]: Adding two spaces in front of this `%if` is possible, but we have to carefully trim whitespaces [using + or - signs](https://jinja.palletsprojects.com/en/2.11.x/templates/#whitespace-control) at the end of the directive and the reader of the code has to understand that writing:
```
{% if ... %}
``` 
is different from
```
  {% if ... -%}
``` 
I would keep this modification as a follow up after v0.16.0, if you still think it's good to have. In any case it would be an internal refactor without changes to user facing behavior.

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: Tried to reword in this sense in 6ccff2a. Let me know if you see other possible improvements.

File: share/spack/templates/container/singularity.def
Comment (scheibelp) [reply share/spack/templates/container/singularity.def]: > I would keep this modification as a follow up after v0.16.0, if you still think it's good to have

I think this is important to have. I'd like to figure out how to address the template whitespace issues before merging this PR.

I think you are saying that to use proper indentation in the template, you will have to add this formatting syntax so that the resulting generated file does not contain the extra whitespace. To clarify my original suggestion in https://github.com/spack/spack/pull/15028#discussion_r451228201, I was saying that for now just indent the directives in the template file and allow the resulting generated files to contain additional whitespace.

I agree that the manual whitespace management syntax looks unpleasant and not worth adding. The documentation you linked also appears to contain a universal setting for removing whitespace associated with template directives (e.g. constructing the `jinja2.Environment` with `trim_blocks=True` and `lstrip_blocks=True`) - is there a reason not to do that?

Comment (scheibelp) [reply share/spack/templates/container/singularity.def]: To be clear, one of my proposed solutions in https://github.com/spack/spack/pull/15028#discussion_r524813712 is hopefully very easy (but results in messier generated files), so if that is acceptable then I think this should be quick to resolve.

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: typo: "unecessary" -> "unnecessary"

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst:233]: This appears to address https://github.com/spack/spack/pull/15028#discussion_r451242531

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: Typo: "its" -> "it's"

File: lib/spack/spack/schema/container.py
Comment (scheibelp) [reply lib/spack/spack/schema/container.py:65]: I'm going to mark this as resolved but to clarify: my suggestion is to let the user specify the full setup/teardown commands e.g. `apt-get install x y z` instead of specifying `apt` and `[x, y, z]`; the former would allow users more flexibility in setting up their container prior to the Spack command.

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: Typo: "use" -> "uses"

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: Suggested wording change: "the user is just requested to specify" -> "all the user has to do is specify"

**Review by scheibelp — CHANGES_REQUESTED** [review]
My main request is to address whitespace handling in the template files (see https://github.com/spack/spack/pull/15028/files#r399576494). The rest of the comments are very minor (generally typos).

File: lib/spack/docs/containers.rst
Comment (scheibelp) [inline lib/spack/docs/containers.rst]: Additional request: I have the same complaint about this comment that I made in https://github.com/spack/spack/pull/15028#discussion_r419749236

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst:233]: Yes, that was the intent.

File: lib/spack/spack/schema/container.py
Comment (alalazo) [reply lib/spack/spack/schema/container.py:65]: That's a good point, and we may need to adjust it later based on user feedback. The current state is less flexible but also less verbose than what you propose. If the lack of flexibility becomes a problem for users we should definitely change it though.

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: :+1: Will fix here and in the template for the Dockerfile

File: lib/spack/docs/containers.rst
Comment (alalazo) [reply lib/spack/docs/containers.rst]: Let me know if the current wording seems better

**Review by scheibelp — APPROVED** [review]

**Pull request merged**
